File: 1544-count-good-nodes-in-binary-tree/count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # you also want to pass max value besides root as args
    # create a new function for that
    # you also want to maintain global count variable
    # the initial value for max is -float("inf") since the n range starts from 1
    def goodNodes(self, root: TreeNode) -> int:
        self.count = 0
        self.dfs(root, -float("inf"))
        return self.count


    def dfs(self, root, max_value):
        if root is None:
            return 
        elif root.val >= max_value:
            self.count += 1
            max_value = root.val
        self.dfs(root.left, max_value)
        self.dfs(root.right, max_value)

File: 1544-count-good-nodes-in-binary-tree/count-good-nodes-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int goodNodes(TreeNode root) {
        return goodNodes(root, -(int) Math.pow(10, 4) -1);
    }

    public int goodNodes(TreeNode root, int maxScore) {
        if (root == null) {
            return 0;
        }
        int count = maxScore <= root.val ? 1 : 0;
        int newMaxScore = Math.max(maxScore, root.val);
        count += goodNodes(root.left, newMaxScore);
        count += goodNodes(root.right, newMaxScore);
        return count;
    }
}